### DISCUSSION TITLE
Tsfresh features for unsupervised clustering

### DISCUSSION THREAD

**Discussion by Sukanya2191** [blue-yonder/tsfresh#861] [Q&A]
Dear All,

I have extracted the features using tsfresh for accelerometer signal data.
Can anyone please help me which clustering algorithm to be used for tsfresh features.

Regards,
Sukanya

**Comment by nils-braun** [comment]
Hi @Sukanya2191 !
It seems no one in the community has any real ideas to share here!
Unfortunately, I can also not really help you. Other than: you will need to try it out (as with all ML methods) and probably do some grid-search. I would not expect there is anything special on clustering tsfresh data in contrast to any other ML clustering problem.

Did you maybe in the meantime already came up with a solution that you want to share?

**Comment by e5k** [comment]
Hi @nils-braun 

I am attempting some clustering and will share the method. In the meanwhile, is there a way to extract relevant features when y is not defined? Just wondering if I am missing something or if y is required to compute relevance. Thanks!

**Comment by nils-braun** [comment]
Hi @e5k !
That would be much appreciated - thanks!

No, it is impossible to extract relevant features without knowing the target. Not because it is not implemented in tsfresh, but because it is not possible: when the target is (yet) unknown, a relevance of the feature is undefined (think about it this way: a feature is relevant for one target, but could be irrelevant for another target. If the target is unknown, how should we know if it is relevant or not?).
That being said, if you only want to know if the features for example describe your data well, you could try to create a forecasting task out of your data. Your target would then be the value at the next (time)step and you could use this for feature selection. However, depending on your real problem/target, that might not help.

**Comment by rpanai** [comment]
In order to clusterize timeseries you could have a look at [pyts](https://github.com/johannfaouzi/pyts) and [tslearn](https://github.com/tslearn-team/tslearn). Behind both these packages there is the same guy @johannfaouzi and in this [paper](https://jmlr.org/papers/volume21/19-763/19-763.pdf) there is an introduction and comparative of several time series packages in python. **tslearn** do the clustering using classical ML algorithms and it looks to me it's applying them directly to the time series without extract features before.

**Comment by e5k** [comment]
@rpanai Thanks a lot for these references Roberto

**Comment by johannfaouzi** [comment]
pyts is focused on time series classification only and does not provide utilities for clustering. tslearn has a [clustering](https://tslearn.readthedocs.io/en/stable/gen_modules/tslearn.clustering.html#module-tslearn.clustering) submodule with several algorithms.

**Comment by nils-braun** [comment]
@e5k Did you had a chance to try out clustering with `tsfresh` or were you using one of the libraries proposed in the other comments?

**Comment by kempa-liehr** [comment]
@e5k We have recently published a paper on unsupervised feature selection, which might be useful for your clustering problem: https://doi.org/10.1109/JSEN.2021.3084970

**Comment by e5k** [comment]
@nils-braun  ashamed I missed your comment! I was using other libraries, then the project took me away from clustering but I might get back to it soon. @kempa-liehr Thanks a lot, will check it out

**Comment by kkorovesis** [comment]
@Sukanya2191 Did you find a solution to your problem ? I have a similar project, I have a list of products with time-series data and I want to detect duplicate products based on there metrics. So I think that by clustering the features created with tsfresh I should expect to find duplicate/similar product in the dataset. Did you find out how to use tsfresh features for cluatering ?
